### PR TITLE
Semconv 1.25.0 migration

### PR DESCRIPTION
**Pull request by SylvainJuge**
The [1.24.0](https://github.com/open-telemetry/semantic-conventions-java/releases/tag/v1.24.0) release of semconv for java brings breaking changes that prevent a simple upgrade PR like https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/10971 to work as expected.

As there are lots of changes I've split the PR in separate commits:
- 1b46cf57d1f438748ce80aa46b4989a3096d9f6a update to semconv 1.24.0 from https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/10971
- f2d90cfa507e98f88b340258ce7844b7acc851e6 introduce a temporary module to keep compile-time compatibility and minimize code changes
- migrate semconv attributes and remove temporary module (about 1 commit per semconv namespace).

For `SemanticAttributes.EXCEPTION_EVENT_NAME` I had to inline and duplicate the `"exception"` value as there isn't any equivalent in the new semantic conventions.

### Follow-up questions where input is welcome
- [x] should we keep a copy of the old semconv classes to make migration easier (https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/10983#pullrequestreview-1976429247).
  Answer is "YES" and this was done
- [x] if yes, should we keep it here in the agent or in a separate repo/artifact ? (https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/10983#issuecomment-2034363531)
  Answer is "NO", it remains as part of the "stable" semconv artifact due to the usage of java modules that prevent having classes from a package spread in more than one artifact.

### Follow-up tasks once this is merged
- https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/11069
- https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/11070

### CODE REVIEW

File: conventions/src/main/kotlin/io.opentelemetry.instrumentation.javaagent-shadowing.gradle.kts [hunk 1]
```diff
@@ -28,6 +28,7 @@ tasks.withType<ShadowJar>().configureEach {
     // relocate(OpenTelemetry API) since these classes live in the bootstrap class loader
     relocate("io.opentelemetry.api", "io.opentelemetry.javaagent.shaded.io.opentelemetry.api")
     relocate("io.opentelemetry.semconv", "io.opentelemetry.javaagent.shaded.io.opentelemetry.semconv")
+    relocate("io.opentelemetry.semconv.incubating", "io.opentelemetry.javaagent.shaded.io.opentelemetry.semconv.incubating")
```
Comment (laurit) [inline conventions/src/main/kotlin/io.opentelemetry.instrumentation.javaagent-shadowing.gradle.kts:31]: could you also update https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/120da53bd30f03f54b7a09313c9e747c1363673a/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/RemappingUrlConnection.java#L35 That is used for loading extensions that refer to these classes with their original name.

**Review by laurit — COMMENTED** [review]
Thanks for working on this.
I think we are going to need to preserve the old semconv classes in the agent to avoid breaking extensions that could be using the old semconv classes. We could do this either by copying these semconv classes like you did in the previous version of this PR or by doing some build script magic. This can be done in a separate PR. @trask WDYT?

File: conventions/src/main/kotlin/io.opentelemetry.instrumentation.javaagent-shadowing.gradle.kts (hunk 1 above)
Comment (SylvainJuge) [reply conventions/src/main/kotlin/io.opentelemetry.instrumentation.javaagent-shadowing.gradle.kts:31]: Fixed in 435d3aeb022a4ddcb3d8efa885db2205dc1c8e8c

**Comment by SylvainJuge** [comment]
> Thanks for working on this. I think we are going to need to preserve the old semconv classes in the agent to avoid breaking extensions that could be using the old semconv classes. We could do this either by copying these semconv classes like you did in the previous version of this PR or by doing some build script magic. This can be done in a separate PR. @trask WDYT?

If every project that depends on semconv java is impacted, then it could make sense to provide such "compatibility layer" directly in https://github.com/open-telemetry/semantic-conventions-java/ as a separate artifact to make migration easier (with of course deprecation warnings), then drop it after a few releases to push migration and avoid having to keep it forever, or we could even include it as part of the `semconv-incubating` artifact to make it obvious that it won't be there forever without having a separate artifact.

File: instrumentation/resources/library/src/test/java/io/opentelemetry/instrumentation/resources/OsResourceTest.java [hunk 2]
```diff
@@ -23,119 +24,119 @@ class OsResourceTest {
   @SetSystemProperty(key = "os.name", value = "Linux 4.11")
   void linux() {
     Resource resource = OsResource.buildResource();
-    assertThat(resource.getSchemaUrl()).isEqualTo(ResourceAttributes.SCHEMA_URL);
-    assertThat(resource.getAttribute(ResourceAttributes.OS_TYPE))
-        .isEqualTo(ResourceAttributes.OsTypeValues.LINUX);
-    assertThat(resource.getAttribute(ResourceAttributes.OS_DESCRIPTION)).isNotEmpty();
+    assertThat(resource.getSchemaUrl()).isEqualTo(SchemaUrls.V1_24_0);
```
Comment (SylvainJuge) [inline instrumentation/resources/library/src/test/java/io/opentelemetry/instrumentation/resources/OsResourceTest.java:27]: [for reviewer] I am not sure what is the "right" value to pick here, so I picked the 1.24.0 to match the semconv version.

**Comment by trask** [comment]
> Thanks for working on this.

💯

> I think we are going to need to preserve the old semconv classes in the agent to avoid breaking extensions that could be using the old semconv classes. We could do this either by copying these semconv classes like you did in the previous version of this PR or by doing some build script magic. This can be done in a separate PR. @trask WDYT?

I'd support copying the old semconv classes in (seems easiest) and marking them all as deprecated. We can remove the copies in 3.0 when we bump database and RPC semconv.

**Review by JonasKunz — APPROVED** [review]

**Comment by trask** [comment]
just holding off on merging this for the next semconv java 1.25.0 release, so we can avoid breaking any Java agent extensions that were relying on the old classes

**Comment by trask** [comment]
semconv java 1.25.0 has been released

@SylvainJuge when you have time, can you update to 1.25.0 and resolve conflicts, then we'll merge it!

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/semconv/network/internal/NetworkAttributes.java [hunk 3]
```diff
@@ -1,29 +0,0 @@
...
-import static io.opentelemetry.api.common.AttributeKey.longKey;
-import static io.opentelemetry.api.common.AttributeKey.stringKey;
-
-import io.opentelemetry.api.common.AttributeKey;
-
-/**
- * This class is internal and is hence not for public use. Its APIs are unstable and can change at
- * any time.
- */
-public final class NetworkAttributes {
```
[the remark this answers is not in the document]
Comment (laurit) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/semconv/network/internal/NetworkAttributes.java:17]: As far as I know this class was temporarily introduced to use attributes that weren't present int the latest semconv jar. It is fine to remove it now.

**Review by laurit — APPROVED** [review]

**Pull request merged**